src/sinks/util/sink.rs

```diff
@@ -325,35 +337,45 @@ where
             let this = self.as_mut().project();
             let mut partitions_ready = vec![];
             for (partition, batch) in this.partitions.iter() {
-                if (*this.closing && !batch.is_empty())
+                if ((*this.closing && !batch.is_empty())
                     || batch.was_full()
                     || matches!(
                         this.lingers
                             .get_mut(partition)
                             .expect("linger should exists for poll_flush")
                             .poll_unpin(cx),
                         Poll::Ready(())
-                    )
+                    ))
+                    && this
+                        .in_flight
+                        .as_mut()
+                        .and_then(|map| map.get_mut(partition))
+                        .map(|req| matches!(req.poll_unpin(cx), Poll::Ready(())))
+                        .unwrap_or(true)
                 {
                     partitions_ready.push(partition.clone());
                 }
             }
             let mut batch_consumed = false;
             for partition in partitions_ready.iter() {
-                let service_ready = match self.service.poll_ready(cx) {
+                let service_ready = match this.service.poll_ready(cx) {
                     Poll::Ready(Ok(())) => true,
                     Poll::Ready(Err(error)) => return Poll::Ready(Err(error)),
                     Poll::Pending => false,
                 };
                 if service_ready {
                     trace!("Service ready; Sending batch.");
 
-                    let batch = self.partitions.remove(partition).unwrap();
-                    self.lingers.remove(partition);
+                    let batch = this.partitions.remove(partition).unwrap();
+                    this.lingers.remove(partition);
 
                     let batch_size = batch.num_items();
                     let (batch, finalizers) = batch.finish();
-                    tokio::spawn(self.service.call(batch, batch_size, finalizers));
+                    let future = tokio::spawn(this.service.call(batch, batch_size, finalizers));
+
+                    this.in_flight.as_mut().map(|map| {
+                        map.insert(partition.clone(), future.map(|_| ()).fuse().boxed())
+                    });
```

Comment:
I'm surprised clippy doesn't want you to change this to `if let Some(map) = this.in_flight.as_mut() {…`

Reply:
It probably gave up.